### ISSUE TITLE
Base layer of ironing top fill is not monotonic

### ISSUE THREAD

**Issue by TheBoojah** [supermerill/Slic3r#1708]
Using ironing as top infill, the ironing pass is monotonic, which is great. However the base infill / first ironing pass on the top layer is not done monotonically. 

Perhaps I do something wrong or have a slight over extrusion, but the ironing pass is not quite able to overcome the overlap in sections from the non-monotonic first pass for me. I suppose the first pass should be under extruded such that non-monotonic does not matter, but for me it seems that is does.  

Please update such that both passes on the top layer with ironing is done monotonically.

Project file:
[PerimeterFanspeed.zip](https://github.com/supermerill/SuperSlicer/files/7418101/PerimeterFanspeed.zip)
Version 2.3.57.2

**Comment by TheBoojah** [comment]
Doing a bunch of calibration and tuning trying to get a great top layer, still have some tests to run before I can say the i/p overlap is doing it.

Did notice another thing, there is a gap next to the gap-fill:
![image](https://user-images.githubusercontent.com/33135012/139946747-a9a0566e-7194-42a2-84aa-66b543864781.png)
![image](https://user-images.githubusercontent.com/33135012/139946920-a6e5e5b6-515f-4a42-b8c6-ddc783c687cc.png)
Closeup:
![image](https://user-images.githubusercontent.com/33135012/139946982-9493c3b1-26c7-4609-956e-2bf69e8f3026.png)
![image](https://user-images.githubusercontent.com/33135012/139949621-b1bea03d-3974-4e58-bb28-9c019aca8380.png)

The ironing passes does not go close enough to the gap fill line, so there is a hole in the final part. This is even though the overall top layer is over-extruded elsewhere.

[Reindeer ironing torture test.zip](https://github.com/supermerill/SuperSlicer/files/7463841/Reindeer.ironing.torture.test.zip)

**Comment by neophyl** [comment]
Isnt this because ironing only irons topfill, which is the red areas so it wouldnt touch the gapfill.  It would need changing to add any gapfill on the toplayer into the ironing path I think.

**Comment by TheBoojah** [comment]
Yeah, ideally the second pass should go over the gap fill, however not on the really thin fills like in the last image. In general gap filled areas are pretty smooth so it's not super necessary I think. But it could bridge the gap between the regular infill and the gap fill better if ironing went everywhere. Also any travel-zits and such will be smoothed over.

But the issue i describe above is that the red 1st pass is actively avoiding the gap fill, producing a gap!

**Comment by supermerill** [comment]
> there is a gap next to the gap-fill:

I know, fixed in 4959f1d

**Comment by TheBoojah** [comment]
> I know, fixed in [4959f1d](https://github.com/supermerill/SuperSlicer/commit/4959f1d29ce9a4c8f9cab89b589863bf95791fda)

Excellent, tried the nightly and it looks much better, thanks!

> You can use the ironing post-process instead of the pattern, it should go over everything.

Good to know! I'll try and see if there are any visible differences on the end result later.


> To disable extraneous gapfill and let infill go here, you have to ensure that `gapfill`->`after last perimeter` isn't checked.

Not sure I understand this option or its tooltip.
After last perimeter checked:
![image](https://user-images.githubusercontent.com/33135012/140033779-9c5cf869-fb83-4314-a994-f3c95d94eb00.png)

After last perimeter NOT checked:
![image](https://user-images.githubusercontent.com/33135012/140033846-22950f67-143d-4685-968b-a5db37496974.png)

Not checked = more gap fill.
You are saying not checked = disable extra gap fill?

**Comment by supermerill** [comment]
> You are saying not checked = disable extra gap fill?

lol, yes... maybe try with rectilinear top file, just to see? I guess that the second gapfill comes from the monotonic_filled algo

**Comment by TheBoojah** [comment]
> maybe try with rectilinear top file, just to check

I tried, the top infill type does not change how that gap fill behaves.

**Comment by supermerill** [comment]
surely becasue of "1 peri on top", so it thinks there is more perimeter incoming...
It may be a bit complicated to solve that one...
